File: LeetCode/P25Hard_Reverse_Nodes_in_k-Group/main.cpp
// Problem: https://leetcode.com/problems/reverse-nodes-in-k-group/
#include <utility>

// Definition for singly-linked list.
struct ListNode {
  int val;
  ListNode *next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
  std::pair<ListNode *, ListNode *> reverse(ListNode *head) {
    ListNode *curr = head, *prev = nullptr;

    while (curr) {
      ListNode *nxt = curr->next;
      curr->next = prev;
      prev = curr;
      curr = nxt;
    }
    return {prev, head};
  }

  ListNode *reverseKGroup(ListNode *head, int k) {
    ListNode sent = ListNode(-1, head), *curr = head, *preced = &sent;

    int c = 1;
    while (curr) {
      if (c == k) {
        ListNode *nxt = curr->next;
        curr->next = nullptr;
        auto [head, tail] = reverse(preced->next);
        tail->next = nxt;
        preced->next = head;
        preced = curr = tail;
        c = 0;
      }

      curr = curr->next;
      c += 1;
    }
    return sent.next;
  }
};

File: LeetCode/P25Hard_Reverse_Nodes_in_k-Group/main.py
from typing import Optional
# Problem: https://leetcode.com/problems/reverse-nodes-in-k-group/


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def reverse(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head:
            return None, None
        prev = None
        curr = head
        while curr:
            nxt = curr.next
            curr.next = prev
            prev = curr
            curr = nxt
        # returns head and tail of new list
        return prev, head

    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        sent = prevk = ListNode(-1, head)
        prev = sent
        curr = head

        while True:
            for i in range(k):
                if not curr:
                    return sent.next
                else:
                    prev = curr
                    curr = curr.next
            prev.next = None
            # reverse list
            revhead, revtail = self.reverse(prevk.next)

            prevk.next = revhead
            revtail.next = curr
            prevk = revtail
